### ISSUE TITLE
Cannot navigate back to home

### ISSUE THREAD

**Issue by ebrehault** [plone/plone.restapi-angular#33]
Thanks to angular-traversal, we can navigate from page to page and then use our browser's Back button to navigate back:

- we are at /folder1
- we navigate to /folder/page1
- we click Back
- we get /folder1

But when navigating back to the site root, we remain on the last page:

- we are at /
- we navigate to /page1
- we click Back
- we are still on /page1

Note: I am not sure if it is an angular-traversal bug or a plone.restapi-angular one.

**Comment by sunew** [comment]
I have a clue to this bug - I think it is caused by the traverser in angular-traversal, traverser.ts lines 48-54 and lines 60-63.

We need to be clear on the specs for the Traverser.
Is it supposed to be able to take partial paths, like '@@view' ?
If so I suggest we write additional testcases to 
clarify this spec.

If not, I do not understand why the possibility in traverse() to work without a contextPath is needed.

I am working on refactoring the path handling (including the handling of slash or no slash etc.) with regards to baseHref and the backend api path (the path to plone), and I suggest using the methods build in to angular Location (or Http) to handle path and querystring manipulation. Perhaps encapsulated in a separate service in angular-traversal.

Another goal for me is trying to make it work to run on the same domain, with either the ng serve proxy use for development or nginx proxying for prod. It mostly works out of the box, but sometimes the path handling with regards to BACKEND_URL gets a little strange (when setting BACKEND_URL to '/site'
 - without domain) 
This would also solve links for images and downloads for files etc. 
And make development across platforms easier (using the same ip/domain for the api as the app is running on in the browser).
More on this later, but what are you initial thoughts? (I should move this to a separate ticket).
I suggest we should handle two cases:
api on separate domain / path
api on same domain as app, with either separate path or with setting up content negotiation - plone.restapi already uses content negotiation as the general principle for serving an uri as json.

**Comment by ebrehault** [comment]
@sunew we definitetly want to support partiel path like `@view`

And your refactoring sounds very good :)

**Comment by ebrehault** [comment]
@sunew by the way, will you be at Barcelona Plone conf next week?

**Comment by sunew** [comment]
@ebrehault Plone Conf: it could be great, also to sprint a bit on these things - working on getting some funding but I think its going to be difficult to make it.

**Comment by ebrehault** [comment]
@sunew too bad...
But we will organise some dedicated sprints regarding restapi/react/angular during the year. It is very likely to have a sprint in Bonn soon. I would be nice if you could join us :)

**Comment by sunew** [comment]
@ebrehault I am going - just made all the ends meet :) See you at the conference.

**Comment by ebrehault** [comment]
Excellent! See you there :)

**Comment by Glamurio** [comment]
Very happy for the reunion! However, back on to the topic at hand, I might be running into this very issue, has there been any progress on the issue?

**Comment by ebrehault** [comment]
@Glamurio could you try with the current angular-traverser master version?

**Comment by Glamurio** [comment]
@ebrehault Yup, just did, still seem to be getting the following error:

```
ERROR TypeError: Object(...) is not a function
    at Traverser.traverse (angular-traversal.js:135)
```

**Comment by ebrehault** [comment]
@Glamurio I have just merged the Angular 7 version (tagged 2.2.0 on master)

**Comment by Glamurio** [comment]
@ebrehault Just tried a fresh installation, still get the following error:

`Module not found: Error: Can't resolve '@angular/http' in '..\plone-angular\node_modules\@plone\restapi-angular\dist\views'`

Also, you still do not seem to have included the rxjs-compat dependency, which requires manual installation, but at least that's easily solvable.

**Comment by ebrehault** [comment]
rxjs-compat is not needed anymore and `@angular/http` is deprecated

Your error traceback shows you are using `@plone\restapi-angular` from `node_modules`, you must use a GitHub checkout to get the latest version.

**Comment by Glamurio** [comment]
Sorry, still kind of new to using GitHub, I'm just used to getting packages via npm, I'll try that and report back if everything works.

**Comment by Glamurio** [comment]
Alright, so I've installed the newest version and I'm running into the issue that I cannot import the module via:
`import { RESTAPIModule } from 'restapi-angular';`

It simply says that the module cannot be found, but on that note, it's probably smarter to create a new issue regarding this and close the current one, as it has nothing to do with the issue anymore.

**Comment by ebrehault** [comment]
I recommend you use mrs-developer https://www.npmjs.com/package/mrs-developer, it will set the proper path in your tsconfig.ts

**Comment by Glamurio** [comment]
Sorry, I've been trying to use it, but for some reason I cannot get it to work? I try to run the command `missdev` in the console and it keeps telling me there's no such command.

Am I using it incorrectly?

**Comment by ebrehault** [comment]
`npm install -g mrs-developer`

**Comment by Glamurio** [comment]
Ah! I'm an idiot, I forgot the `-g` tag for it. Thanks, it works now, I'll figure out the rest.

**Comment by ebrehault** [comment]
:)

**Comment by Glamurio** [comment]
Actually, hm, it seems I'm running into yet another issue, when using missdev, it gives me an error:


`Error: spawn git ENOENT`
`    at Process.ChildProcess._handle.onexit (internal/child_process.js:264:19)`
`    at onErrorNT (internal/child_process.js:456:16)`
`    at processTicksAndRejections (internal/process/task_queues.js:84:9)`

**Comment by Glamurio** [comment]
Okay, I tried around for a while now, but no matter how I try to use missdev, it always gives me the following error: 

```
Using src\develop
(node:16976) UnhandledPromiseRejectionWarning: ReferenceError: err is not defined
    at ...\AppData\Roaming\npm\node_modules\mrs-developer\src\index.js:107:70
    at processTicksAndRejections (internal/process/task_queues.js:89:5)
    at async develop (...\AppData\Roaming\npm\node_modules\mrs-developer\src\index.js:139:21)
(node:16976) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 1)
(node:16976) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
```

**Comment by ebrehault** [comment]
oh it seems you are on Windows, right?

Never tried on Windows, no idea if it works. mrs-developer depends on simple-git which just need the `git` command to be available in the PATH, so that's maybe the problem here.

I have just published mrs-developer 1.1.3, which display the error properly, so you will know more about what is wrong.

**Comment by Glamurio** [comment]
I'd gotten an error regarding git which was resolved after installing git, but after just now updating to the newest version, the error I get is:

```
Using src\develop
Cannot open src\develop\plone.restapi-angular No repo
Cannot checkout plone.restapi-angular
Update paths in tsconfig.json
```

**Comment by ebrehault** [comment]
ok well, just checkout manually plone.restapi-angular in src\develop and you'll be fine

**Comment by Glamurio** [comment]
I'm sorry, what does checkout mean exactly? Like I said, I'm still new to all this, I'm not entirely sure what to do.

**Comment by ebrehault** [comment]
Just that:
```
cd src/develop
git clone https://github.com/plone/plone.restapi-angular.git
```
:)

**Comment by Glamurio** [comment]
Thanks a lot! That worked. It now seems like it correctly updated the paths according to mrs developer, however my VSC still tells me it cannot find the module. While trying to serve for instance, I get this error:

```
ERROR in src/app/app.module.ts(3,31): error TS2307: Cannot find module '@plone/restapi-angular'.
src/develop/plone.restapi-angular/src/app/app.module.ts(7,73): error TS2307: Cannot find module 'ngx-schema-form'.
```

This error made me check up on 'ngx-schema-form' which it was trying to read from `src/develop/plone.restapi-angular/src/app/app.module.ts`, so I went there and checked the `tsconfig.json` as well as the `mrs.developer.json` to see that it was also trying to read it from a develop folder, which naturally didn't exist. 

Thinking I could fix it by simply replicating what I did earlier with the git checkout however only yielded even more cryptic errors. Any ideas?

**Comment by ebrehault** [comment]
yes exactly, checkout https://github.com/guillotinaweb/ngx-schema-form.git in src/develop too and but the proper path in tsconfig.json:
```
"ngx-schema-form": [
                "develop/ngx-schema-form/projects/schema-form/src/lib"
            ],
```

You might also need angular-traversal https://github.com/guillotinaweb/angular-traversal.git:
```
"angular-traversal": [
                "develop/angular-traversal/projects/traversal/src/lib"
            ],
```

**Comment by Glamurio** [comment]
Well, as I mentioned, the issue comes from the fact that it cannot read `ngx-schema-form` from the `app.module.ts` inside the `plone.restapi-angular`. Which means in order to fix that I have to do a git checkout inside of `plone.restapi-angular`, but if I do that, I get the following error instead:
```

ERROR in ../app/app.module.ts(3,31): error TS2307: Cannot find module '@plone/restapi-angular'.
ngx-schema-form/projects/schema-form/src/lib/formelement.action.component.ts(27,50): error TS2345: Argument of type '{ read: typeof ViewContainerRef; static: boolean; }' is not assignable to parameter of type '{ read?: any; }'.
  Object literal may only specify known properties, and 'static' does not exist in type '{ read?: any; }'.
ngx-schema-form/projects/schema-form/src/lib/widgetchooser.component.ts(29,50): error TS2345: Argument of type '{ read: typeof ViewContainerRef; static: boolean; }' is not assignable to parameter of type '{ read?: any; }'.
  Object literal may only specify known properties, and 'static' does not exist in type '{ read?: any; }'.
ngx-schema-form/src/typings.d.ts(3,18): error TS2300: Duplicate identifier 'value'.
plone.restapi-angular/src/develop/ngx-schema-form/projects/schema-form/src/lib/formelement.action.component.ts(27,50): error TS2345: Argument of type '{ read: typeof ViewContainerRef; static: boolean; }' is not assignable to parameter of type '{ read?: any; }'.
  Object literal may only specify known properties, and 'static' does not exist in type '{ read?: any; }'.
plone.restapi-angular/src/develop/ngx-schema-form/projects/schema-form/src/lib/widgetchooser.component.ts(29,50): error TS2345: Argument of type '{ read: typeof ViewContainerRef; static: boolean; }' is not assignable to parameter of type '{ read?: any; }'.
  Object literal may only specify known properties, and 'static' does not exist in type '{ read?: any; }'.
plone.restapi-angular/src/develop/ngx-schema-form/src/typings.d.ts(3,18): error TS2300: Duplicate identifier 'value'.
```

**Comment by ebrehault** [comment]
If ngx-schema-form is declared in your main tsconfig.json file, it will be resolved where it is mentionned.

**Comment by Glamurio** [comment]
Okay, this seems to be the case. However I still get the exact same error as described above. Something about types not being assignable to `type '{ read?: any; }'`.

**Comment by ebrehault** [comment]
oh that's probably because you are not using angular 8

either you use master branches with angular 8, either you checkout earlier tags for angular 7 or 6

**Comment by Glamurio** [comment]
Okay, I've been using Angular 7 on the last project to test all this functionality but since you mentioned Angular 8 compatibility I've started a fresh new project with Angular 8. After following every step, it all worked, however I had to additionally do a git checkout for "z-schema", but since I knew how it worked now, that was no problem.

The `ng build --prod` works without any issues as well, so not only am I able to run it on Angular 8 now, but I'm also able to build properly. Thanks a lot for all your assistance, I don't think I'd have been able to figure this out by myself! I kept wanting to tell you to close this issue and open a new one, but it was easier to just chat in here, but I think it's safe to say we can close this issue now, and if there are any more problems, I'll open a new one. c:
